### ISSUE TITLE
Process pooling / re-using the same Puppeteer instance for multiple screenshots?

### ISSUE THREAD

**Issue by olalonde** [sindresorhus/pageres#284]
Is there any way to re-use the same phantomjs instance for multiple screenshots if the list of URLs aren't known in advance?

**Comment by SamVerschueren** [comment]
duplicate of #242

**Comment by olalonde** [comment]
It's not really, unless the answer is that "no" this won't be implemented. What I'm looking for is to have a "single" phantomjs across all screenshots and this is not possible at the moment.

**Comment by SamVerschueren** [comment]
It was requested here though https://github.com/sindresorhus/pageres/issues/282#issuecomment-252893771. 

As @kevva points out, we don't think it would be possible to change the URL while keeping the process alive.

**Comment by olalonde** [comment]
I've actually just started working on that using `node-phantomjs`: https://gist.github.com/a1f2d6f56db7d8bbf0f8616ab5018346 (WIP!)

It re-uses the same phatomjs `instance` (process) for all screenshots within the same Node.js process. It would require a pretty big refactoring of pageres as far as I can tell, so I wrote it from scratch. I'd be happy to contribute to pageres but the way it's designed doesn't make it easy.

**Comment by SamVerschueren** [comment]
I don't think it's worth the effort right now. Once headless chrome lands (not sure when it will happen though), we will switch to that one. So then we could take a look at how it could be done in headless chrome.

**Comment by olalonde** [comment]
Ok, sure. FWIW, after switching from pageres to the code I just [pasted](https://gist.github.com/anonymous/a1f2d6f56db7d8bbf0f8616ab5018346), I got a huge performance boost (possibly 10x) in my script:

(each migration generates a screenshot from a url)

Before:
```
info: Migrating 5eVMdE
info: Migrated 5eVMdE in 8.70s
info: Migrating mep2bM
info: Migrated mep2bM in 8.60s
info: Migrating 7ax9by
info: Migrated 7ax9by in 8.47s
info: Migrating YaOpbx
info: Migrated YaOpbx in 7.52s
info: Migrating RdG7dD
info: Migrated RdG7dD in 8.56s
info: Migrating QdJ2dO
info: Migrated QdJ2dO in 8.94s
info: Migrating 1aKReQ
```

After:
```
info: Migrating RdG7dD
info: Migrated RdG7dD in 5.94s
info: Migrating QdJ2dO
info: Migrated QdJ2dO in 1.19s
info: Migrating mbk5ez
info: Migrated mbk5ez in 1.21s
info: Migrating 7ax9by
info: Migrated 7ax9by in 1.13s
info: Migrating 1aKReQ
info: Migrated 1aKReQ in 1.19s
info: Migrating wdLwbj
info: Migrated wdLwbj in 1.20s
info: Migrating 1aMAaW
```

So I think it's worth considering in the future, for use cases where performance matters.

**Comment by olalonde** [comment]
Phantom resource pool: https://gist.github.com/88c0d6cfcd410d655d53eaa787f6b317

I might end up open sourcing this, haha  👍  feel free to re-use

**Comment by olalonde** [comment]
Released this under https://github.com/blockai/phantom-pool

**Comment by sindresorhus** [comment]
Decided not to do this.
